sphinx/builders/html.py

```diff
@@ -404,15 +404,22 @@ def get_doc_context(self, docname, body, metatags):
         # title rendered as HTML
         title = self.env.longtitles.get(docname)
         title = title and self.render_partial(title)['title'] or ''
+
+        # Suffix for the document
+        source_suffix = path.splitext(self.env.doc2path(docname))[1]
+
         # the name for the copied source
-        sourcename = self.config.html_copy_source and docname + '.txt' or ''
+        if self.config.html_copy_source:
+            if self.config.html_sourcelink_keep_suffix:
+                sourcename = docname + source_suffix
+            else:
+                sourcename = docname + '.txt'
```

Comment:
What about `html_sourcelink_append_txt = lambda self: self.source_suffix != '.txt'`

@ericholscher Would this break anything in RTD?